### ISSUE TITLE
NPE hardware acceleration support in Qualcomm chips

### ISSUE THREAD

**Issue by madhavajay** [tensorflow/tensorflow#17526]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: N/A
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Android OS
- **TensorFlow installed from (source or binary)**: N/A
- **TensorFlow version (use command below)**: 1.6 TFLite
- **Python version**:  N/A
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: See below
- **Exact command to reproduce**: N/A

Can anyone clarify what chips will support NPE and have hardware acceleration for on device inference with tf lite?

NPE SDK says 845, 820, 835, 625, 626, 650, 652, 653, 660, 630, 636, 450, 820Am.

I was under the impression that only the 845 had NPE?

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by madhavajay** [comment]
Yes, it has received no answer and I have reached out on Twitter and in the Qualcomm forums and no one has answered what the relationship is between TensorFlow Lite, NNAPI, sNPE and Qualcomm chips.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by madhavajay** [comment]
I understand some details might have been tight lipped before the TF 2018 Dev Summit, but I gather from the talks that the Qualcomm Hexagon DSP driver support for NNAPI will come in Android P. Theres a lot of talk in the Dev Summit about things like NVIDIA TensorRT and Intel MKL-DNN etc as well as the ability to deploy models to platforms like Mobile and Raspberri Pi, so my guess is that these hardware specific optimisations will be supported through some higher level TensorFlow graph optimisation tools and for Android that will be NNAPI. So I guess the Hexagon Chip support will be up to Qualcomms provided NNAPI driver in Android P, but my guess is Flagships like 845, and then possibly the same line up above.

Any word on the Pixel devices? Seems kind of cruel that the Google Pixel's 821 had the Hexagon DSP drivers disabled for sNPE?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by llhe** [comment]
@madhavajay 
For running neural nets on Qualcomm chips, you can have look at [MACE](https://github.com/xiaomi/mace), which is highly optimized for mobile phone performance. Here are some [benchmark results](https://github.com/XiaoMi/mace/issues/1).

**Comment by madhavajay** [comment]
Wow thats really impressive stuff. Thanks for the notification! 👍

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rockyrhodes** [comment]
SNPE is Qualcomm's neural processing SDK for their chips.  It provides support for running TF-trained models on various Qualcomm Snapdragon SoC chips.  The SDK information page lists a long list of supported chips, as pointed out by @madhavajay.  Stock android, however, doesn't provide support for the underlying communication mechanism between the CPU and the DSP, so HVX acceleration via SNPE isn't supported on those platforms.  Android P, running on SDM845-based platforms, will accelerate inferences on the HVX via Qualcomm's NNAPI HAL driver.  I'm not certain about plans for NNAPI HAL support on platforms based on other Qualcomm chips, but this is definitely not the same thing as SNPE support for the underlying SoC chip.

Sorry for the complicated answer.

**Comment by madhavajay** [comment]
@rockyrhodes thanks for the reply. Yes, I guess another way to phrase it would be, when will Android and TensorFlow support acceleration on Qualcomm chips that leverages the Hexagon DSP. It was mentioned in the TF Summit 2018 that it would come in Android P.

**Comment by reuben** [comment]
@rockyrhodes sorry for replying to a closed issue, but just to clarify, will Android P on MSM8998 (Snapdragon 835) get HVX acceleration? It has the Hexagon 682 versus the 685 on the SDM845. Thanks!

**Comment by rockyrhodes** [comment]
@reuben: I have no information about Qualcomm's plans for NNAPI HAL drivers for other chips, sorry.

**Comment by UnsGentoals** [comment]
Check this.
https://developer.qualcomm.com/docs/snpe/overview.html
